### ISSUE TITLE
Buffer overflow in pngfix

### ISSUE THREAD

**Issue by patacca** [glennrp/libpng#350]
There is a buffer overflow inside contrib/tools/pngfix.c
```c
int
main(int argc, const char **argv)
{
   char temp_name[FILENAME_MAX+1];

...

         size_t outlen = strlen(*argv);

...

            else if (suffix != NULL)
               memcpy(temp_name, *argv, outlen);

            temp_name[outlen] = 0;

            if (suffix != NULL)
...
```

temp_name buffer has a dimension of FILENAME_MAX+1 and outlen can be arbitrary specified.

In order to trigger the buffer overflow we have to supply the option --suffix=xxx and a filename greater than FILENAME_MAX+1 (typically 4096+1)

for example
```bash
$ ./pngfix --suffix=xxx $(python -c 'print("A"*5000)')
```

**Comment by jbowler** [comment]
This seems to be a duplicate of #294, albeit a lot better bug report since it seems to correctly explain the problem.

**Comment by jbowler** [comment]
The fix seems reasonable and correct.

**Comment by ctruta** [comment]
Fixed in commit 8a5732fcb30b8afc4d3c23144acf2b502bb80122 and integrated in libpng-1.6.39. Many thanks for the defect report, for the fix, and for the code review.
